tech-specs/src/specs/ledger.md

```diff
@@ -0,0 +1,190 @@
+1. Attempt to call the module's entrypoint function. The entrypoint MUST have signature:
+   ```wat
+   func (param i64 i64 i64 i64 i64 i64 i64 i64) (result i64))
+   ```
+   - The first argument is the offset to the owner’s address written into the memory, the second argument is its bytes length
+   - The third is the offset of the transaction’s `data` and fourth is it’s bytes length
+   - The fifth is the offset of the modified storage keys and sixth is its bytes length
+   - The seventh is the offset of the triggered validity predicates owners' addresses and eighth is its bytes length
+
+#### Gas
+
+##### Gas constants
+
+The gas constants are currently chosen arbitrarily and are subject to change following gas accounting estimations.
+
+| Name                   | Value |
+|------------------------|-------|
+| `COMPILE_GAS_PER_BYTE` |     1 |
+| `BASE_TRANSACTION_FEE` |     2 |
+| `PARALLEL_GAS_DIVIDER` |    10 |
+| `MIN_STORAGE_GAS`      |     1 |
+
+- TODO describe gas accounting, wasm gas counter, limits, what happens if we go over limits and how gas relates to fees
+
+#### WebAssembly (WASM)
+
+##### WASM constants
+
+| Name                                 | Unit              | Value |
+|--------------------------------------|-------------------|-------|
+| `PAGE` (as defined in the WASM spec) | kiB               |    64 |
+| `TX_MEMORY_INIT_PAGES`               | number of `PAGE`s |   100 |
+| `TX_MEMORY_MAX_PAGES`                | number of `PAGE`s |   200 |
+| `VP_MEMORY_INIT_PAGES`               | number of `PAGE`s |   100 |
+| `VP_MEMORY_MAX_PAGES`                | number of `PAGE`s |   200 |
+| `WASM_STACK_LIMIT`                   | stack depth       | 65535 |
+
+
+The WASM instantiation, the types, instructions, validation and execution of WASM modules MUST conform to the [WebAssembly specification](https://webassembly.github.io/spec/core/intro/index.html).
+
+##### WASM validation
+
+The WebAssembly code is REQUIRED to only use deterministic instructions. Furthermore, it MUST NOT use features from any of the following WebAssembly proposals:
+
+- The reference types proposal
+- The multi-value proposal
+- The bulk memory operations proposal
+- The module linking proposal
+- The SIMD proposal
+- The threads proposal
+- The tail-call proposal
+- The multi memory proposal
+- The exception handling proposal
+- The memory64 proposal
+
+##### Stack height limiter
+
+To make stack overflows deterministic, set the upper bound of the stack size to [`WASM_STACK_LIMIT`](#wasm-constants). If the stack height exceeds the limit then execution MUST abort.
+
+<!--
+cargo test test_tx_stack_limiter
+cargo test test_vp_stack_limiter
+-->
+
+##### WASM memory
+
+- TODO memory read/write gas costs
+
+##### Transaction host environment functions
+
+The following functions from the host ledger are made available in transaction's WASM code. They MAY be imported in the WASM module as shown bellow and MUST be provided by the ledger's WASM runtime:
```

Comment:
Can we explain a bit about what each of these functions do?

Reply:
same as below, left a TODO placeholder